File: Arkane/src/Channel.cpp
#include "Channel.h"
#include "Sample.h"

Channel::Channel(void)
{
	sample1 = new Sample();
	highPassLevel=lowPassLevel=0;
	dspLowPass = 0;
    dspHighPass = 0;
	dspflange=0;
	pan=0;
	volume=1;
	isMuted = false;
	
}
Channel::~Channel(void)
{

}
bool Channel::Initialize(FMOD::System* fmodSystemPtr, int index)
{
	channelIndex = index;
	fmodSystem = fmodSystemPtr;
	fmodSystem->getChannel(channelIndex, &fmodChannel);
	//-----LowPass
	fmodSystem->createDSPByType(FMOD_DSP_TYPE_LOWPASS, &dspLowPass);
	fmodSystem->addDSP(dspLowPass, 0);
	dspLowPass->setParameter(FMOD_DSP_LOWPASS_CUTOFF,22000);
	dspLowPass->getActive(&dsplowpass_active);
	//------HighPass
	fmodSystem->createDSPByType(FMOD_DSP_TYPE_HIGHPASS, &dspHighPass);
	fmodSystem->addDSP(dspHighPass, 0);
	dspHighPass->setParameter(FMOD_DSP_HIGHPASS_CUTOFF,10);
	dspHighPass->getActive(&dsphighpass_active);
	return true;
}
FMOD_RESULT Channel::LoadSample(char* fileName)
{
	FMOD_RESULT result = FMOD_OK;
	sample1->CreateSampleFromFile(fileName, fmodSystem);
	return result;
}
FMOD_RESULT Channel::PlaySample(){
	sample1->Play(fmodSystem, fmodChannel);
	SetVolume(volume);
	return FMOD_OK;
}
void Channel::SetVolume(float newVol)
{
	if(!isMuted){
		FMOD_RESULT result = fmodChannel->setVolume(newVol);
	}

}
float Channel::GetVolume()
{
	float vol;
	FMOD_RESULT result = fmodChannel->getVolume(&vol);
	return vol;
}
void Channel::SetPlaying(bool playing)
{
	isPlaying = playing;
}
bool Channel::IsMuted(){
	if(isMuted)return true;
	else return false;
}
bool Channel::IsPlaying(){
	if(isPlaying)return true;
	else return false;
}
void Channel::Mute(){
	volumeBeforeMuting=GetVolume();
	SetVolume(0.0f);
	isMuted=true;
}
void Channel::UnMute()
{
	isMuted=false;
}
void Channel::SetReverb(){
	FMOD::DSP *dsphighpass   = 0;
	FMOD_RESULT result = fmodSystem->createDSPByType(FMOD_DSP_TYPE_HIGHPASS, &dsphighpass);
	result = fmodSystem->addDSP(dsphighpass, 0);
}
void Channel::SetHighPass(float input)
{
	//highPassLevel=input;
	highPassLevel=input*220+7.8;
	
	if(highPassLevel<10)highPassLevel=10;
	else if(highPassLevel>22000)highPassLevel=22000;
	dspHighPass->setParameter(FMOD_DSP_HIGHPASS_CUTOFF,highPassLevel);	
}
float Channel::GetHighPass()
{
	return highPassLevel;	
}

void Channel::SetLowPass(float input)
{
	input = 100 - input;
	lowPassLevel=input*220+7.8;
	
	if(lowPassLevel<10)lowPassLevel=10;
	else if(lowPassLevel>22000)lowPassLevel=22000;

	dspLowPass->setParameter(FMOD_DSP_LOWPASS_CUTOFF,lowPassLevel);
	
}
float Channel::GetLowPass()
{
	return lowPassLevel;
}
void Channel::SetPanLevel(float newPanLevel)
{
	pan=newPanLevel;
	fmodChannel->setPan(newPanLevel);
}
float Channel::GetPanLevel()
{
	fmodChannel->getPan(&pan);
	return pan;
}

File: CSharp/Channel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using FMOD;

namespace arkane
{
    public class ArkaneChannel
    {
        FMOD.Channel channel;
        int channelIndex;
        float pan = 0;
        bool isActive = false;
        bool isPlaying = false;
        bool isMuted = false;
        bool isPaused = false;
        uint currentTime;
        float volumeBeforeMuting;
        Sample sample1;
        bool active;
        bool dsphighpass_active;
        bool dsplowpass_active;
        float highPassLevel, lowPassLevel = 0;
        FMOD.DSP dspLowPass, dspHighPass, dspFlange;
        public FMOD.RESULT result;
        FMOD.System fmodSystem;
        DSPConnection con = new DSPConnection();
        public FMOD.System system;

        public ArkaneChannel(FMOD.System sys)
        {
            this.fmodSystem = sys;
            this.sample1 = new Sample(fmodSystem);
        }

        public void Initialize(int index)
        {
            channelIndex = index;
            fmodSystem.getChannel(channelIndex, ref channel);
            // -------Low Pass
            fmodSystem.createDSPByType(DSP_TYPE.LOWPASS, ref dspLowPass);
            fmodSystem.addDSP(dspLowPass, ref con);
            dspLowPass.setParameter((int)DSP_LOWPASS.CUTOFF, 0);
            dspLowPass.getActive(ref dsplowpass_active);

            // --------High Pass
            fmodSystem.createDSPByType(DSP_TYPE.HIGHPASS, ref dspHighPass);
            fmodSystem.addDSP(dspHighPass, ref con);
            dspHighPass.setParameter((int)DSP_LOWPASS.CUTOFF, 0);
            dspHighPass.getActive(ref dsphighpass_active);

        }

        public void LoadSample(String fileName)
        {
            sample1.CreateSampleFromFile(fileName);
        }

        public void PlaySample()
        {
            if (isPaused)
            {
                sample1.Play(channel, currentTime);
                SetVolume(1);
                isPlaying = true;
                isPaused = false;
            }
            else
            {
                sample1.Play(channel);
                SetVolume(1);
                isPlaying = true;
                isPaused = false;
            }
        }

        public void SetVolume(float newVol)
        {
            if (!isMuted)
            {
                result = channel.setVolume(newVol);
                Sample.ERRCHECK(result);
            }
        }

        public float GetVolume()
        {
            float vol = -1;
            result = channel.getVolume(ref vol); Sample.ERRCHECK(result);
            return vol;
        }

        public void SetPlaying(bool playing)
        {
            isPlaying = playing;
        }

        public bool IsPlaying()
        {
            return isPlaying;
        }

        public void Mute()
        {
            volumeBeforeMuting = GetVolume();
            SetVolume(0.0f);
            isMuted = true;
        }

        public void SetReverb()
        {
            DSP dsphighpass = null;
            result = fmodSystem.createDSPByType(DSP_TYPE.HIGHPASS, ref dsphighpass);
            Sample.ERRCHECK(result);
            result = fmodSystem.addDSP(dsphighpass, ref con);
            Sample.ERRCHECK(result);
        }

        public void SetHighPass(float input)
        {
            // I don't like magic numbers
            highPassLevel = (float)(input * 220 + 7.8);

            if (highPassLevel < 10) highPassLevel = 10;
            else if (highPassLevel > 22000) highPassLevel = 22000;
            result = dspHighPass.setParameter((int)DSP_HIGHPASS.CUTOFF, highPassLevel);
            Sample.ERRCHECK(result);
        }

        public float GetHighPass()
        {
            return highPassLevel;
        }

        public void SetLowPass(float input)
        {
            input = 100 - input;
            lowPassLevel = (float)(input * 220 + 7.8);
            
            if (lowPassLevel < 10) lowPassLevel = 10;
            else if (lowPassLevel > 22000) lowPassLevel = 22000;

            result = dspLowPass.setParameter((int)DSP_LOWPASS.CUTOFF, lowPassLevel);
            Sample.ERRCHECK(result);
        }

        public float GetLowPass()
        {
            return lowPassLevel;
        }

        public void SetPanLevel(float newPanLevel)
        {
            pan = newPanLevel;
            result = channel.setPan(newPanLevel);
            Sample.ERRCHECK(result);
        }

        public float GetPanLevel()
        {
            result = channel.getPan(ref pan);
            Sample.ERRCHECK(result);
            return pan;
        }

        public void PlaySoundAt(uint start, uint duration) 
        {
            currentTime = start;
            sample1.Play(this, start, duration);
        }
        
        public void SetLoop(bool isLoop) 
        {
            if (isLoop) result = channel.setMode(MODE.LOOP_NORMAL);
            else result = channel.setMode(MODE.LOOP_OFF);
            Sample.ERRCHECK(result);
        }
        public void StopSample() 
        {
            result = channel.stop(); Sample.ERRCHECK(result);
            isPlaying = false;
        }

        public uint GetPosition()
        {
            uint time = 0;
            channel.getPosition(ref time, TIMEUNIT.BUFFERED);
            return time;
        }

        public void Pause()
        {
            this.StopSample();
            this.isPaused = true;
            result = channel.getPosition(ref this.currentTime, TIMEUNIT.BUFFERED);
            Sample.ERRCHECK(result);
        }
    }
}
